profiles/car.lua

```diff
@@ -287,328 +279,444 @@ function way_function (way, result)
+      data.result.backward_speed = data.result.backward_speed * scaling_factor
     end
   end
+end
 
-  local is_bidirectional = result.forward_mode ~= mode.inaccessible and result.backward_mode ~= mode.inaccessible
-
-  -- scale speeds to get better avg driving times
-  if result.forward_speed > 0 then
-    local scaled_speed = result.forward_speed * speed_reduction
-    local penalized_speed = math.huge
-    if service and service ~= "" and service_speeds[service] then
-      penalized_speed = service_speeds[service]
-    elseif width <= 3 or (lanes <= 1 and is_bidirectional) then
-      penalized_speed = result.forward_speed / 2
-    end
-    result.forward_speed = math.min(penalized_speed, scaled_speed)
-  end
+-- determine if this way can be used as a start/end point for routing
+function handle_startpoint(data)
+  -- only allow this road as start point if it not a ferry
+  data.result.is_startpoint = data.result.forward_mode == mode.driving or 
+                              data.result.backward_mode == mode.driving
+end
 
-  if result.backward_speed > 0 then
-    local scaled_speed = result.backward_speed * speed_reduction
-    local penalized_speed = math.huge
-    if service and service ~= "" and service_speeds[service]then
-      penalized_speed = service_speeds[service]
-    elseif width <= 3 or (lanes <= 1 and is_bidirectional) then
-      penalized_speed = result.backward_speed / 2
-    end
-    result.backward_speed = math.min(penalized_speed, scaled_speed)
-  end
+-- leave early if this way is not accessible
+function handle_way_type(data)
+  if "" == TagCache.get(data,"highway") then return false end
+end
 
-  -- Handle high frequency reversible oneways (think traffic signal controlled, changing direction every 15 minutes).
-  -- Scaling speed to take average waiting time into account plus some more for start / stop.
-  if oneway and "alternating" == oneway then
-    local scaling_factor = 0.4
-    if result.forward_speed ~= math.huge then
-      result.forward_speed = result.forward_speed * scaling_factor
-    end
-    if result.backward_speed ~= math.huge then
-      result.backward_speed = result.backward_speed * scaling_factor
-    end
-  end
+-- set the road classification based on guidance globals configuration
+function handle_classification(data)
+  set_classification(TagCache.get(data,"highway"),data.result,data.way)
+end
 
-  -- only allow this road as start point if it not a ferry
-  result.is_startpoint = result.forward_mode == mode.driving or result.backward_mode == mode.driving
+-- main entry point for processsing a way
+function way_function(way, result)
+  
+  -- define a table that we can pass around to helper functions
+  -- so they have access to the input/output objects.
+  
+  -- we also use a table as a local cache of tags, to avoid calling
+  -- into C++ more than once for each tag, without having to
+  -- pass lists of already fetches tags around.
+
+  local data = {
```

Comment:
This might be more of a style thing, but I think passing these members unwrapped to the helper functions like `handle_initial_check(way)` and `handle_speed(way, result)` make it more explicit what is actually used.